itemloaders/__init__.py

```diff
@@ -414,10 +419,57 @@ def get_css(self, css, *processors, **kw):
             # HTML snippet: <p id="price">the price is $1200</p>
             loader.get_css('p#price', TakeFirst(), re='the price is (.*)')
         """
-        values = self._get_cssvalues(css, **kw)
+        values = self.get_selector_values(None, css, 'css', **kw)
         return self.get_value(values, *processors, **kw)
 
-    def _get_cssvalues(self, csss, **kw):
+    def get_selector_values(self, field_name, selector_rules, selector_type, **kw):
+
         self._check_selector_method()
-        csss = arg_to_iter(csss)
-        return flatten(self.selector.css(css).getall() for css in csss)
+
+        selector = getattr(self.selector, selector_type or '', None)
+
+        # The optional arg in methods like `add_css()` for context in stats
+        name = kw.get("name")
+
+        # For every call of `add_css()` and `add_xpath()` this is incremented.
+        # We'll use it as the base index of the position of the logged stats.
+        index = self.field_tracker[f"{field_name}_{selector_type}"]
+
+        values = []
+        for position, rule in enumerate(arg_to_iter(selector_rules), index):
+            parsed_data = selector(rule).getall()
+            values.append(parsed_data)
+            self.write_to_stats(
+                field_name, parsed_data, position, selector_type, name=name
+            )
+        return flatten(values)
+
+    def write_to_stats(
+        self, field_name, parsed_data, position, selector_type, name=None
+    ):
+        """Responsible for logging the parser rules usage.
+
+        The implementation below where each missing parsed_data is being logged
+        to the stat is clunky, but necessary. With this, we can only surmise
+        that it's safe to remove parser fallback parser if it's all just
+        '.../missing' in the stats.
+        """
+
+        if not self.stats or not field_name:
+            return
+
+        parser_label = (
+            f"parser/{self.loader_name}/{field_name}/{selector_type}/{position}"
+        )
+
+        if name:
+            parser_label += f"/{name}"
+
+        if parsed_data in (None, []):
+            parser_label += "/missing"
+
+        self.stats.inc_value(parser_label)
```

Comment:
This makes itemloaders depend on scrapy.stats, at least as far as interface is concerned. Do you think we can avoid it? I can see 2 main ways:

* ask for a different stats object here, maybe a plain dictionary; allow to pass real stats in Scrapy's itemloader
* move stats handling to Scrapy's wrapper instead of this package

We can also start decoupling scrapy Stats from scrapy, but I'd rather not do this now.

Reply:
Hi @kmike , the itemloaders accept Scrapy's stats instance via DI and could work without it, [as per this](https://github.com/scrapy/itemloaders/pull/12/files/17139441f8199a2fdf96eb73c27c0619d634502c#diff-2873a0f550b5bf030b1bb1545b09c842R458).

Accepting the actual Scrapy stats instance is crucial for the loader's usage to be logged at the end of a job (_and be logged in Scrapy Cloud's stats-tab_), like this:

```
2019-06-16 14:32:32 [scrapy.statscollectors] INFO: Dumping Scrapy stats:
{ ...
  'parser/QuotesItemLoader/author/css/1': 10,
  'parser/QuotesItemLoader/quote/css/1/missing': 10,
  'parser/QuotesItemLoader/quote/css/2': 10
  ...
}
```

I'm all good for making it like a `defaultdict(int)` to function like the stat's `inc_value`, as long as we can pass it out to Scrapy's logging system.

However, I might be missing something? 🤔Is there another way to circumvent injecting Scrapy's stats to the itemloaders and still log the parser's usage?

Reply:
I guess @kmike 's point here is the implicit dependency.
Even though, scrapy is not a requirement, we require a direct dependency from scrapy..
So, maybe a solution would be to create some sort of adapter (another one).
Though, I somehow foresee a problem.

If we allow `ItemLoader` to be extended with `upkeep`, we'll need to make `upkeep` the default base class for scrapy, as scrapy users we'll keep using scrapy's `ItemLoader` which is an extension to this one.
So, if they import this extension instead of scrapy's, they'll be losing scrapy's behavior.